hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/DoubleGaugeImplTest.java

```diff
@@ -113,4 +146,186 @@ public void whenDoubleGaugeField() {
+
+        someObject.doubleField = 42.65D;
+        assertEquals(42.65D, doubleGauge.read(), 10E-6);
+
+    }
+
+    @Test
+    public void whenCreatedForDynamicLongMetricWithExtractedValue() {
+        SomeObject someObject = new SomeObject();
+        someObject.longField = 42;
+        metricsRegistry.registerDynamicMetricsProvider(someObject);
+        DoubleGauge doubleGauge = metricsRegistry.newDoubleGauge("foo.longField");
+
+        // needed to collect dynamic metrics and update the gauge created from them
+        metricsRegistry.collect(mock(MetricsCollector.class));
+
+        assertEquals(42D, doubleGauge.read(), 10E-6);
+
+        someObject.longField = 43;
+        assertEquals(43D, doubleGauge.read(), 10E-6);
+    }
+
+    @Test
+    public void whenCreatedForDynamicLongMetricWithProvidedValue() {
+        DoubleGaugeImplTest.SomeObject someObject = new DoubleGaugeImplTest.SomeObject();
+        someObject.longField = 42;
+        metricsRegistry.registerDynamicMetricsProvider(
+                (taggerSupplier, context) -> context
+                        .collect(taggerSupplier.getMetricTagger("foo"), "longField", INFO, BYTES, 42));
+        DoubleGauge doubleGauge = metricsRegistry.newDoubleGauge("foo.longField");
+
+        // needed to collect dynamic metrics and update the gauge created from them
+        metricsRegistry.collect(mock(MetricsCollector.class));
+
+        assertEquals(42, doubleGauge.read(), 10E-6);
+    }
+
+    @Test
+    public void whenCreatedForDynamicDoubleMetricWithProvidedValue() {
+        DoubleGaugeImplTest.SomeObject someObject = new DoubleGaugeImplTest.SomeObject();
+        someObject.longField = 42;
+        metricsRegistry.registerDynamicMetricsProvider(
+                (taggerSupplier, context) -> context
+                        .collect(taggerSupplier.getMetricTagger("foo"), "doubleField", INFO, BYTES, 41.65D));
+        DoubleGauge doubleGauge = metricsRegistry.newDoubleGauge("foo.doubleField");
+
+        // needed to collect dynamic metrics and update the gauge created from them
+        metricsRegistry.collect(mock(MetricsCollector.class));
+
+        assertEquals(41.65, doubleGauge.read(), 10E-6);
+    }
+
+    @Test
+    public void whenCacheDynamicMetricSourceGcdReadsDefault() {
+        NullableDynamicMetricsProvider metricsProvider = new NullableDynamicMetricsProvider();
+        WeakReference<SomeObject> someObjectWeakRef = new WeakReference<>(metricsProvider.someObject);
+        metricsProvider.someObject.doubleField = 42.42D;
+        metricsRegistry.registerDynamicMetricsProvider(metricsProvider);
+        DoubleGauge doubleGauge = metricsRegistry.newDoubleGauge("foo.doubleField");
+
+        // needed to collect dynamic metrics and update the gauge created from them
+        metricsRegistry.collect(mock(MetricsCollector.class));
+        assertEquals(42.42D, doubleGauge.read(), 10E-6);
+
+        metricsProvider.someObject = null;
+        System.gc();
```

Comment:
`System.gc()` is a hint for JVM and I am not sure we have to rely on it in the unit test. Otherwise, we risk to get a flaky test.